File: app/src/test/java/bj/discogsbrowser/login/LoginPresenterTest.java
package bj.discogsbrowser.login;

import com.github.scribejava.core.model.OAuth1AccessToken;
import com.github.scribejava.core.oauth.OAuth10aService;

import org.fuckboilerplate.rx_social_connect.Response;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.junit.MockitoJUnitRunner;

import bj.discogsbrowser.utils.SharedPrefsManager;
import bj.discogsbrowser.wrappers.RxSocialConnectWrapper;
import io.reactivex.Observable;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

/**
 * Created by Josh Laird on 09/05/2017.
 */
@RunWith(MockitoJUnitRunner.class)
public class LoginPresenterTest
{
    private LoginPresenter presenter;
    @Mock LoginContract.View view;
    @Mock SharedPrefsManager sharedPrefsManager;
    @Mock RxSocialConnectWrapper rxSocialConnectWrapper;
    @Mock OAuth10aService oAuth10aService;
    private OAuthTokenFactory oAuthTokenFactory = new OAuthTokenFactory();

    @Before
    public void setUp()
    {
        presenter = new LoginPresenter(view, sharedPrefsManager, rxSocialConnectWrapper, oAuth10aService);
    }

    @After
    public void tearDown()
    {
        verifyNoMoreInteractions(view, sharedPrefsManager, oAuth10aService);
    }

    @Test
    public void startOAuthServiceValid_logsIn()
    {
        LoginActivity mockLoginActivity = mock(LoginActivity.class);
        when(view.getActivity()).thenReturn(mockLoginActivity);
        when(rxSocialConnectWrapper.with(mockLoginActivity, oAuth10aService)).thenReturn(Observable.just(new Response<>(mockLoginActivity, oAuthTokenFactory.getValidToken())));

        presenter.beginLogin();

        verify(view).getActivity();
        verify(sharedPrefsManager).storeOAuthToken(any(OAuth1AccessToken.class));
        verify(view).finishActivityLaunchMain();
    }

    @Test
    public void startOAuthServiceError_viewDisplaysError()
    {
        LoginActivity mockLoginActivity = mock(LoginActivity.class);
        when(view.getActivity()).thenReturn(mockLoginActivity);
        when(rxSocialConnectWrapper.with(mockLoginActivity, oAuth10aService)).thenReturn(Observable.error(new Throwable()));

        presenter.beginLogin();

        verify(view).getActivity();
        verify(view).displayErrorDialog();
    }
}

File: app/src/test/java/bj/vinylbrowser/login/LoginPresenterTest.kt
package bj.vinylbrowser.login

import bj.vinylbrowser.utils.SharedPrefsManager
import bj.vinylbrowser.wrappers.RxSocialConnectWrapper
import com.github.scribejava.core.model.OAuth1AccessToken
import com.github.scribejava.core.oauth.OAuth10aService
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.verify
import com.nhaarman.mockito_kotlin.verifyNoMoreInteractions
import com.nhaarman.mockito_kotlin.whenever
import io.reactivex.Observable
import org.fuckboilerplate.rx_social_connect.Response
import org.junit.After
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.ArgumentMatchers.any
import org.mockito.junit.MockitoJUnitRunner

/**
 * Created by Josh Laird on 21/05/2017.
 */
@RunWith(MockitoJUnitRunner::class)
class LoginPresenterTest {
    lateinit var presenter: LoginPresenter
    val view: LoginContract.View = mock()
    val sharedPrefsManager: SharedPrefsManager = mock()
    val rxSocialConnectWrapper: RxSocialConnectWrapper = mock()
    val oAuth10aService: OAuth10aService = mock()

    @Before
    fun setUp() {
        presenter = LoginPresenter(view, sharedPrefsManager, rxSocialConnectWrapper, oAuth10aService)
    }

    @After
    fun tearDown() {
        verifyNoMoreInteractions(view, sharedPrefsManager, oAuth10aService)
    }

    @Test
    fun startOAuthServiceValid_logsIn() {
        val mockLoginActivity: LoginActivity = mock()
        whenever(view.activity).thenReturn(mockLoginActivity)
        whenever(rxSocialConnectWrapper.with(mockLoginActivity, oAuth10aService)).thenReturn(Observable.just(Response<LoginActivity, OAuth1AccessToken>(mockLoginActivity, OAuthTokenFactory.getValidToken())))

        presenter.beginLogin()

        verify(view).activity
        verify(sharedPrefsManager).storeOAuthToken(any(OAuth1AccessToken::class.java))
        verify(view).finishActivityLaunchMain()
    }

    @Test
    fun startOAuthServiceError_viewDisplaysError() {
        val mockLoginActivity: LoginActivity = mock()
        whenever(view.activity).thenReturn(mockLoginActivity)
        whenever(rxSocialConnectWrapper.with(mockLoginActivity, oAuth10aService)).thenReturn(Observable.error<Response<LoginActivity, OAuth1AccessToken>>(Throwable()))

        presenter.beginLogin()

        verify(view).activity
        verify(view).displayErrorDialog()
    }
}